### PR TITLE
Add support for testconfig.json entries in MSTest

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/3572

```

// Expected format of the json is: -
//
// "mstest" : {
//  "timeout" : {
//      "assemblyInitialize" : strictly positive int,
//      "assemblyCleanup" : strictly positive int,
//      "classInitialize" : strictly positive int,
//      "classCleanup" : strictly positive int,
//      "testInitialize" : strictly positive int,
//      "testCleanup" : strictly positive int,
//      "test" : strictly positive int
 //      "useCooperativeCancellation" : true/false
//  },
//  "parallelism" : {
//      "enabled": true/false,
//      "workers": positive int,
//      "scope": method/class,
//  },
//  "execution": {
//     "collectSourceInformation": true,
//     "executionApartmentState": "STA",
//     "disableAppDomain": true,
//     "mapInconclusiveToFailed" : true,
//     "mapNotRunnableToFailed" : true,
//     "treatDiscoveryWarningsAsErrors" : true,
//     "considerEmptyDataSourceAsInconclusive" : true,
//     "treatClassAndAssemblyCleanupWarningsAsErrors" : true,
//     "considerFixturesAsSpecialTests" : true,
//  },
//  "deployment" : {
//       "enabled": true / false,
//       "deployTestSourceDependencies": true / false,
//       "deleteDeploymentDirectoryAfterTestRunIsComplete": true / false
//  },
//  "assemblyResolution" : [
//        { "path" : "..." , includeSubDirectories: "true" } ,
//        { "path" : "..." , includeSubDirectories: "true" } ,
//        { "path" : "..." , includeSubDirectories: "true" } ,
//        ...
//     ],
//  ... remaining settings
// }
```

### CODE REVIEW

File: src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestDiscoverer.cs [hunk 1]
```diff
@@ -17,6 +18,8 @@ namespace Microsoft.VisualStudio.TestPlatform.MSTest.TestAdapter;
 [FileExtension(".exe")]
 public class MSTestDiscoverer : ITestDiscoverer
 {
+    internal static IConfiguration? Configuration { get; private set; }
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestDiscoverer.cs]: is the config read outside or inside a possible appdomain?

File: src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestDiscoverer.cs [hunk 2]
```diff
@@ -26,12 +29,15 @@ public class MSTestDiscoverer : ITestDiscoverer
...
-    public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink)
+    public void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink) => MSTestDiscoverer.DiscoverTests(sources, discoveryContext, logger, discoverySink, null);
+
+    internal static void DiscoverTests(IEnumerable<string> sources, IDiscoveryContext discoveryContext, IMessageLogger logger, ITestCaseDiscoverySink discoverySink, IConfiguration? configuration)
     {
         ValidateArg.NotNull(sources, "sources");
         ValidateArg.NotNull(logger, "logger");
         ValidateArg.NotNull(discoverySink, "discoverySink");
 
+        Configuration = configuration;
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestDiscoverer.cs]: is it static needed? can't we flow the `IConfiguration` down?

Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/VSTestAdapter/MSTestDiscoverer.cs]: I just see it could be long from three diff places.. could it have bad effect?

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 3]
```diff
@@ -265,6 +266,11 @@ public static void PopulateSettings(MSTestSettings settings)
     internal static void PopulateSettings(IDiscoveryContext? context, IMessageLogger? logger)
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: I would expect to see the `IConfiguration?` as part of the method signature.

**Comment by engyebrahim** [comment]
we have that by default generated 
image

so as runsettings file could have some other settings more than mstest should i change the check to search on the key 'key' and throw on the case of it has it and we has it in configration but other than tat we okay two have two files!

**Comment by engyebrahim** [comment]
i didn't use MSTESTv2 key
and from the json example that Amaury added in the task i didn't use that
`"deployment" : {
       "enabled": true/false,
        "deployTestSourceDependencies": true/false
       ,"deleteDeploymentDirectoryAfterTestRunIsComplete": true/false		
       }`
as i didn't find in mstesSettings any deployment settings

**Comment by engyebrahim** [comment]
Directory working i  need more changes from vstest

**Comment by engyebrahim** [comment]
check also visual studio default setting

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 4]
```diff
@@ -262,32 +264,49 @@ public static void PopulateSettings(MSTestSettings settings)
...
-            CurrentSettings = new MSTestSettings();
-            return;
+            throw new InvalidOperationException(Resource.DuplicateConfigurationError);
         }
 
-        MSTestSettings? aliasSettings = GetSettings(context.RunSettings.SettingsXml, SettingsNameAlias, logger);
-
-        // If a user specifies MSTestV2 in the runsettings, then prefer that over the v1 settings.
-        if (aliasSettings != null)
+        if (context?.RunSettings != null && !StringEx.IsNullOrEmpty(context.RunSettings.SettingsXml))
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: Is it the `!StringEx.IsNullOrEmpty(context.RunSettings.SettingsXml)` needed? 
If `context?.RunSettings != null` isn't enough?

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 5]
```diff
@@ -262,32 +264,49 @@ public static void PopulateSettings(MSTestSettings settings)
     /// <param name="logger"> The logger for messages. </param>
     /// The discovery context that contains the runsettings.
     /// </param>
-    internal static void PopulateSettings(IDiscoveryContext? context, IMessageLogger? logger)
+    internal static void PopulateSettings(IDiscoveryContext? context, IMessageLogger? logger, IConfiguration? configuration)
     {
-        RunConfigurationSettings = RunConfigurationSettings.PopulateSettings(context);
-
-        if (context?.RunSettings == null || StringEx.IsNullOrEmpty(context.RunSettings.SettingsXml))
+        if (configuration != null && context?.RunSettings != null)
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: We already discussed this one related to the possible `RunConfig` usage by MSTest.

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 6]
```diff
@@ -262,32 +264,49 @@ public static void PopulateSettings(MSTestSettings settings)
...
 
             CurrentSettings = settings ?? new MSTestSettings();
+
+            return;
         }
 
-        SetGlobalSettings(context.RunSettings.SettingsXml, CurrentSettings, logger);
+        // This will contain default adapter settings
+        CurrentSettings = new MSTestSettings();
+        return;
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: not needed right?

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 7]
```diff
@@ -831,4 +850,162 @@ private static void SetGlobalSettings(
...
+        //  "parallelism" : {
+        //      "enabled": true/false,
+        //      "workers": positive int,
+        //      "scope": method/class,
+        //  },
+        //  ... remaining settings
+        // }
+        MSTestSettings settings = new();
+
+        ParseBooleanSetting(configuration, "captureTraceOutput", logger, result => settings.CaptureDebugTraces = result);
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: I would rename `result` into `value` looks to me clearer what you think?

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 8]
```diff
@@ -831,4 +850,162 @@ private static void SetGlobalSettings(
...
+            else
+            {
+                throw new AdapterSettingsException(string.Format(
+                    CultureInfo.CurrentCulture,
+                    Resource.InvalidClassCleanupLifecycleValue,
+                    classCleanupLifecycle,
+#if NET
+                    string.Join(", ", Enum.GetNames<ClassCleanupBehavior>())));
+#else
+                    string.Join(", ", EnumPolyfill.GetNames<ClassCleanupBehavior>())));
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs:984]: Can't we use polyfill for both? @Evangelink how's the story of polyfill here?

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 9]
```diff
@@ -831,4 +850,162 @@ private static void SetGlobalSettings(
...
+#endif
+            }
+        }
+
+        if (configuration["mstest:parallelism:workers"] is string workers)
+        {
+            settings.ParallelizationWorkers = int.TryParse(workers, out int parallelWorkers)
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs:991]: wonder if we should warning a message in case the value is specified but actually is not a good int to parse.
Fine go on with the default proc count, but looks an important settings that users want to be correct and it's impacting the runtime.

File: src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs [hunk 10]
```diff
@@ -155,4 +159,39 @@ private static RunConfigurationSettings ToSettings(XmlReader reader)
...
+        // }
+
+        // Initialize settings object
+        RunConfigurationSettings settings = new();
+
+        if (bool.TryParse(configuration["mstest:runConfiguration:collectSourceInformation"], out bool collectSourceInformation))
+        {
+            settings.CollectSourceInformation = collectSourceInformation;
+            PlatformServiceProvider.Instance.AdapterTraceLogger.LogInfo(
+                "CollectSourceInformation value Found : {0}", collectSourceInformation);
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs]:
```suggestion
                "CollectSourceInformation value found : {0}", collectSourceInformation);
```

File: src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs [hunk 11]
```diff
@@ -155,4 +159,39 @@ private static RunConfigurationSettings ToSettings(XmlReader reader)
...
+        }
+
+        string? apartmentStateValue = configuration["mstest:runConfiguration:executionApartmentState"];
+        if (!string.IsNullOrEmpty(apartmentStateValue) &&
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs]: I think that the `Enum.TryParse` is enough, I suppose that if you pass a null to it will return `false`.

File: src/Adapter/MSTestAdapter.PlatformServices/Services/SettingsProvider.cs [hunk 12]
```diff
@@ -40,6 +41,15 @@ public static MSTestAdapterSettings Settings
     public static void Reset() => s_settings = null;
 #endif
 
+    internal static void Load(IConfiguration configuration)
+    {
+#if !WINDOWS_UWP
+#pragma warning disable IDE0022 // Use expression body for method
```
Comment (MarcoRossignoli) [inline src/Adapter/MSTestAdapter.PlatformServices/Services/SettingsProvider.cs:47]: I think that this pragma can be removed or?

**Review by MarcoRossignoli — COMMENTED** [review]
Some question genereally looks fine

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs (hunk 4 above)
Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: it was like that in old code, so i think it was for a reason

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs (hunk 6 above)
Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: it's needed if nor runsettings or config files are there o set the default settings

File: src/Adapter/MSTestAdapter.PlatformServices/Services/SettingsProvider.cs (hunk 12 above)
Comment (engyebrahim) [reply src/Adapter/MSTestAdapter.PlatformServices/Services/SettingsProvider.cs:47]: when i remove it it request an arrow function and when i do it I get another error because of `#if`

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs (hunk 9 above)
Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/MSTestSettings.cs:991]: yup we are already throwing in case of invalid value

**Comment by engyebrahim** [comment]
needs to add/fet spacific settings from confic that related to some frameworks like appdomain

File: src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs (hunk 11 above)
Comment (engyebrahim) [reply src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs:184]: when i remove it some tests fails

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 13]
```diff
@@ -831,4 +906,162 @@ private static void SetGlobalSettings(
...
+        //      "workers": positive int,
+        //      "scope": method/class,
+        //  },
+        //  ... remaining settings
+        // }
+        MSTestSettings settings = new();
+
+        ParseBooleanSetting(configuration, "captureTraceOutput", logger, value => settings.CaptureDebugTraces = value);
+        ParseBooleanSetting(configuration, "enableBaseClassTestMethodsFromOtherAssemblies", logger, value => settings.EnableBaseClassTestMethodsFromOtherAssemblies = value);
+        ParseBooleanSetting(configuration, "forcedLegacyMode", logger, value => settings.ForcedLegacyMode = value);
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: Let's not support this setting.

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 14]
```diff
@@ -831,4 +906,162 @@ private static void SetGlobalSettings(
...
+        ParseBooleanSetting(configuration, "mapInconclusiveToFailed", logger, value => settings.MapInconclusiveToFailed = value);
+        ParseBooleanSetting(configuration, "mapNotRunnableToFailed", logger, value => settings.MapNotRunnableToFailed = value);
+        ParseBooleanSetting(configuration, "treatDiscoveryWarningsAsErrors", logger, value => settings.TreatDiscoveryWarningsAsErrors = value);
+        ParseBooleanSetting(configuration, "considerEmptyDataSourceAsInconclusive", logger, value => settings.ConsiderEmptyDataSourceAsInconclusive = value);
+        ParseBooleanSetting(configuration, "treatClassAndAssemblyCleanupWarningsAsErrors", logger, value => settings.TreatClassAndAssemblyCleanupWarningsAsErrors = value);
+        ParseBooleanSetting(configuration, "considerFixturesAsSpecialTests", logger, value => settings.ConsiderFixturesAsSpecialTests = value);
+        ParseBooleanSetting(configuration, "cooperativeCancellationTimeout", logger, value => settings.CooperativeCancellationTimeout = value);
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: This should be part of the timeout section and let's use `useCooperativeCancellation`

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 15]
```diff
@@ -831,4 +906,162 @@ private static void SetGlobalSettings(
...
+
+        ParseIntegerSetting(configuration, "timeout:test", logger, value => settings.TestTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:assemblyCleanup", logger, value => settings.AssemblyCleanupTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:assemblyInitialize", logger, value => settings.AssemblyInitializeTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:classInitialize", logger, value => settings.ClassInitializeTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:classCleanup", logger, value => settings.ClassCleanupTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:testInitialize", logger, value => settings.TestInitializeTimeout = value);
+        ParseIntegerSetting(configuration, "timeout:testCleanup", logger, value => settings.TestCleanupTimeout = value);
+
+        ParseStringSetting(configuration, "settingsFile", logger, fileName => settings.TestSettingsFile = fileName);
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: I think we also don't want to support this one. We are in the new world so there is no point in allowing to load an old setting file

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 16]
```diff
@@ -253,7 +257,61 @@ public static void PopulateSettings(MSTestSettings settings)
     /// The discovery context that contains the runsettings.
     /// </param>
     [Obsolete("this function will be removed in v4.0.0")]
-    public static void PopulateSettings(IDiscoveryContext? context) => PopulateSettings(context, null);
+    public static void PopulateSettings(IDiscoveryContext? context) => PopulateSettings(context, null, null);
+
+    private static bool IsRunSettingsFileHesMSTestSettings(string? runSettingsXml) => IsRunSettingsFileHasSettingName(runSettingsXml, SettingsName, out _) || IsRunSettingsFileHasSettingName(runSettingsXml, SettingsNameAlias, out _)
+                    || (IsRunSettingsFileHasSettingName(runSettingsXml, ConfigurationName, out XmlReader? runConfigrationReader) && HasRelevantKeys(runConfigrationReader));
+
+    private static bool HasRelevantKeys(XmlReader? reader)
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: This seems to only have sense in the cotnext above so I would make it local function of the previous or rename it to something more menaingfull

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 17]
```diff
@@ -831,4 +906,162 @@ private static void SetGlobalSettings(
...
+        //  "parallelism" : {
+        //      "enabled": true/false,
+        //      "workers": positive int,
+        //      "scope": method/class,
+        //  },
+        //  ... remaining settings
+        // }
+        MSTestSettings settings = new();
+
+        ParseBooleanSetting(configuration, "captureTraceOutput", logger, value => settings.CaptureDebugTraces = value);
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: I would have it under some `output` section.

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 18]
```diff
@@ -831,4 +906,162 @@ private static void SetGlobalSettings(
...
+        ParseBooleanSetting(configuration, "enableBaseClassTestMethodsFromOtherAssemblies", logger, value => settings.EnableBaseClassTestMethodsFromOtherAssemblies = value);
+        ParseBooleanSetting(configuration, "forcedLegacyMode", logger, value => settings.ForcedLegacyMode = value);
+        ParseBooleanSetting(configuration, "mapInconclusiveToFailed", logger, value => settings.MapInconclusiveToFailed = value);
+        ParseBooleanSetting(configuration, "mapNotRunnableToFailed", logger, value => settings.MapNotRunnableToFailed = value);
+        ParseBooleanSetting(configuration, "treatDiscoveryWarningsAsErrors", logger, value => settings.TreatDiscoveryWarningsAsErrors = value);
+        ParseBooleanSetting(configuration, "considerEmptyDataSourceAsInconclusive", logger, value => settings.ConsiderEmptyDataSourceAsInconclusive = value);
+        ParseBooleanSetting(configuration, "treatClassAndAssemblyCleanupWarningsAsErrors", logger, value => settings.TreatClassAndAssemblyCleanupWarningsAsErrors = value);
+        ParseBooleanSetting(configuration, "considerFixturesAsSpecialTests", logger, value => settings.ConsiderFixturesAsSpecialTests = value);
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: Let's group them under some `execution` category as suggested by @nohwnd

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs (hunk 8 above)
Comment (Evangelink) [reply src/Adapter/MSTest.TestAdapter/MSTestSettings.cs:984]: I would need to double check but yeah most likely it should work.

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 19]
```diff
@@ -262,32 +290,46 @@ public static void PopulateSettings(MSTestSettings settings)
     /// <param name="logger"> The logger for messages. </param>
     /// The discovery context that contains the runsettings.
     /// </param>
-    internal static void PopulateSettings(IDiscoveryContext? context, IMessageLogger? logger)
+    internal static void PopulateSettings(IDiscoveryContext? context, IMessageLogger? logger, IConfiguration? configuration)
     {
-        RunConfigurationSettings = RunConfigurationSettings.PopulateSettings(context);
-
-        if (context?.RunSettings == null || StringEx.IsNullOrEmpty(context.RunSettings.SettingsXml))
+        if (configuration != null && context?.RunSettings != null && IsRunSettingsFileHasMSTestSettings(context?.RunSettings.SettingsXml))
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]:
```suggestion
        if (configuration != null && context?.RunSettings != null && IsRunSettingsFileHasMSTestSettings(context.RunSettings.SettingsXml))
```

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs (hunk 4 above)
Comment (Evangelink) [reply src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: Simplification looks fine to me:

```suggestion
        if (!StringEx.IsNullOrEmpty(context?.RunSettings.SettingsXml))
```

File: src/Adapter/MSTest.TestAdapter/MSTestSettings.cs [hunk 20]
```diff
@@ -831,4 +873,155 @@ private static void SetGlobalSettings(
...
+        {
+            if (bool.TryParse(value, out bool result))
+            {
+                setSetting(result);
+            }
+            else
+            {
+                logger?.SendMessage(TestMessageLevel.Warning, string.Format(CultureInfo.CurrentCulture, Resource.InvalidValue, value, key));
+            }
+        }
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/MSTestSettings.cs]: Nit: I find it easier mentally when there is less indenting

```suggestion
        if (configuration[$"mstest:{key}"] is not string value)
        {
            return;
        }
        
        if (bool.TryParse(value, out bool result))
        {
            setSetting(result);
        }
        else
        {
            logger?.SendMessage(TestMessageLevel.Warning, string.Format(CultureInfo.CurrentCulture, Resource.InvalidValue, value, key));
        }
```

File: src/Adapter/MSTest.TestAdapter/Resources/Resource.resx [hunk 21]
```diff
@@ -414,4 +414,7 @@ but received {4} argument(s), with types '{5}'.</value>
   <data name="DynamicDataShouldBeValidMessageFormat_MemberType" xml:space="preserve">
     <value>'[DynamicData]' referenced member '{0}.{1}' should return 'IEnumerable&lt;object[]&gt;', 'IEnumerable&lt;Tuple&gt;` or 'IEnumerable&lt;ValueTuple&gt;'</value>
   </data>
+  <data name="DuplicateConfigurationError" xml:space="preserve">
+    <value>You are using both 'runsettings' and 'testconfig' files. Please use only one type of configuration file.</value>
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/Resources/Resource.resx]: Both '.runsettings' and '.testconfig.json' files have been detected. Please select only one of these test configuration files.

File: src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs [hunk 22]
```diff
@@ -155,4 +157,34 @@ private static RunConfigurationSettings ToSettings(XmlReader reader)
...
+        }
+
+        string? apartmentStateValue = configuration["mstest:execution:executionApartmentState"];
+        if (Enum.TryParse(apartmentStateValue, out PlatformApartmentState platformApartmentState))
+        {
+            settings.ExecutionApartmentState = platformApartmentState switch
+            {
+                PlatformApartmentState.STA => ApartmentState.STA,
+                PlatformApartmentState.MTA => ApartmentState.MTA,
+                _ => throw new NotSupportedException($"Platform apartment state '{platformApartmentState}' is not supported."),
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/RunConfigurationSettings.cs]: Exception should be localized, please reuse the exception we have when reading the runsettings.

File: src/Adapter/MSTest.TestAdapter/TestingPlatformAdapter/MSTestBridgedTestFramework.cs [hunk 23]
```diff
@@ -8,17 +8,24 @@
...
 {
+    private readonly IConfiguration? _configration;
+
     public MSTestBridgedTestFramework(MSTestExtension mstestExtension, Func<IEnumerable<Assembly>> getTestAssemblies,
         IServiceProvider serviceProvider, ITestFrameworkCapabilities capabilities)
         : base(mstestExtension, getTestAssemblies, serviceProvider, capabilities)
     {
+        // check if it have mstest settings
+        IConfiguration config = serviceProvider.GetConfiguration();
+        _configration = config["mstest"] is null ? null : config;
```
Comment (Evangelink) [inline src/Adapter/MSTest.TestAdapter/TestingPlatformAdapter/MSTestBridgedTestFramework.cs]: Isn't it better to always pass down the file and let the settings class detect if `mstest` section exists?

File: src/Adapter/MSTestAdapter.PlatformServices/MSTestAdapter.PlatformServices.csproj [hunk 24]
```diff
@@ -31,6 +31,7 @@
 
   <ItemGroup>
     <ProjectReference Include="$(RepoRoot)src\TestFramework\TestFramework.Extensions\TestFramework.Extensions.csproj" />
+    <ProjectReference Include="$(RepoRoot)src\Platform\Microsoft.Testing.Extensions.VSTestBridge\Microsoft.Testing.Extensions.VSTestBridge.csproj" />
```
Comment (Evangelink) [inline src/Adapter/MSTestAdapter.PlatformServices/MSTestAdapter.PlatformServices.csproj:34]: Why is this needed?

File: src/Adapter/MSTestAdapter.PlatformServices/Services/MSTestAdapterSettings.cs [hunk 25]
```diff
@@ -304,5 +360,32 @@ private void ReadAssemblyResolutionPath(XmlReader reader)
...
+        // "mstest" : {
+        //    "assemblyResolution" : [
+        //        { "path" : "..." , includeSubDirectories: "true" } ,
+        //        { "path" : "..." , includeSubDirectories: "true" } ,
+        //        { "path" : "..." , includeSubDirectories: "true" } ,
+        //        ...
+        //     ]
+        //  ... remaining settings
+        // }
+        int indx = 0;
```
Comment (Evangelink) [inline src/Adapter/MSTestAdapter.PlatformServices/Services/MSTestAdapterSettings.cs]: Let's use the full name `index`.

File: test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ConfigurationSettingsTests.cs [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+    [ArgumentsProvider(nameof(TargetFrameworks.All), typeof(TargetFrameworks))]
+    public async Task TestConfigJson_AndRunSettingsWithoutMstest_OverrideRunConfigration(string tfm)
+    {
+        var testHost = TestHost.LocateFrom(_testAssetFixture.ProjectPath, TestAssetFixture.ProjectName, tfm);
+        TestHostResult testHostResult = await testHost.ExecuteAsync("--settings my.runsettings");
+
+        // Assert
+        testHostResult.AssertExitCodeIs(ExitCodes.Success);
+    }
+
```
Comment (Evangelink) [inline test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ConfigurationSettingsTests.cs:49]: Maybe add a last test with no runsettings provided.

**Review by Evangelink — APPROVED** [review]

File: src/Adapter/MSTestAdapter.PlatformServices/MSTestAdapter.PlatformServices.csproj (hunk 24 above)
Comment (engyebrahim) [reply src/Adapter/MSTestAdapter.PlatformServices/MSTestAdapter.PlatformServices.csproj:34]: to use the Iconfiguration from the bridge

**Review by Evangelink — APPROVED** [review]
